File: frontend/src/types/workflow.ts
export type WorkflowPhase = string;
export type WorkflowPhaseStatus =
  | "in_progress"
  | "awaiting_approval"
  | "approved"
  | "revision_requested"
  | "completed";
export type ArtifactStatus = "draft" | "review" | "approved" | "superseded";
export type AnnotationType = "comment" | "suggestion" | "rejection";
export type AnnotationStatus = "pending" | "resolved" | "dismissed";

// ── Workflow Step (노드 속성 인라인) ────────────────────────

export interface WorkflowStepConfig {
  name: string;
  label: string;
  icon: string;
  prompt_template: string;
  constraints: string;
  order_index: number;
  review_required: boolean;
}

// 하위 호환 별칭
export type ResolvedWorkflowStep = WorkflowStepConfig;

// ── Workflow Definition ─────────────────────────────────

export interface WorkflowDefinitionInfo {
  id: string;
  name: string;
  description: string | null;
  is_builtin: boolean;
  is_default: boolean;
  sort_order: number;
  steps: WorkflowStepConfig[];
  created_at: string;
  updated_at: string;
}

export interface ArtifactAnnotationInfo {
  id: number;
  artifact_id: number;
  line_start: number;
  line_end: number | null;
  content: string;
  annotation_type: AnnotationType;
  status: AnnotationStatus;
  created_at: string;
}

export interface SessionArtifactInfo {
  id: number;
  session_id: string;
  phase: string;
  title: string;
  content: string;
  status: ArtifactStatus;
  version: number;
  parent_artifact_id: number | null;
  annotations: ArtifactAnnotationInfo[];
  created_at: string;
  updated_at: string;
}

export interface StartWorkflowRequest {
  workflow_definition_id?: string;
  start_from_step?: string;
  skip_research?: boolean;
  skip_plan?: boolean;
}

export interface WorkflowStatusResponse {
  workflow_enabled: boolean;
  workflow_phase: string | null;
  workflow_phase_status: WorkflowPhaseStatus | null;
  workflow_definition_id: string | null;
  steps: WorkflowStepConfig[];
  artifacts: SessionArtifactInfo[];
}

export interface AddAnnotationRequest {
  line_start: number;
  line_end?: number | null;
  content: string;
  annotation_type?: AnnotationType;
}

export interface UpdateAnnotationRequest {
  status: "resolved" | "dismissed";
}

export interface UpdateArtifactRequest {
  content: string;
}

export interface ApprovePhaseRequest {
  feedback?: string | null;
}

export interface RequestRevisionRequest {
  feedback: string;
}

export interface CreateWorkflowDefinitionRequest {
  name: string;
  description?: string;
  steps: WorkflowStepConfig[];
}

export interface UpdateWorkflowDefinitionRequest {
  name?: string;
  description?: string;
  steps?: WorkflowStepConfig[];
}

export interface WorkflowDefinitionExport {
  version: number;
  definition: WorkflowDefinitionInfo;
}

// ── QA Checklist Types ──────────────────────────────────

export type QAStatus = "pass" | "fail" | "warn";

export interface QAChecklistItem {
  item: string;
  status: QAStatus;
  detail: string;
}

export interface QAChecklistResult {
  all_passed: boolean;
  items: QAChecklistItem[];
  summary: { pass: number; fail: number; warn: number };
}

File: backend/app/schemas/workflow.py
"""워크플로우 관련 Pydantic 스키마."""

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field

WorkflowPhase = Literal["research", "plan", "implement"]
WorkflowPhaseStatus = Literal["in_progress", "awaiting_approval", "approved", "revision_requested"]
ArtifactStatusType = Literal["draft", "review", "approved", "superseded"]
AnnotationType = Literal["comment", "suggestion", "rejection"]
AnnotationStatusType = Literal["pending", "resolved", "dismissed"]


class ArtifactAnnotationInfo(BaseModel):
    """아티팩트 인라인 주석 정보."""

    id: int
    artifact_id: int
    line_start: int
    line_end: Optional[int] = None
    content: str
    annotation_type: AnnotationType
    status: AnnotationStatusType
    created_at: datetime


class SessionArtifactInfo(BaseModel):
    """세션 아티팩트 정보."""

    id: int
    session_id: str
    phase: WorkflowPhase
    title: str
    content: str
    status: ArtifactStatusType
    version: int
    parent_artifact_id: Optional[int] = None
    annotations: list[ArtifactAnnotationInfo] = []
    created_at: datetime
    updated_at: datetime


class StartWorkflowRequest(BaseModel):
    """워크플로우 시작 요청."""

    skip_research: bool = False
    skip_plan: bool = False


class WorkflowStatusResponse(BaseModel):
    """워크플로우 상태 응답."""

    workflow_enabled: bool
    workflow_phase: Optional[WorkflowPhase] = None
    workflow_phase_status: Optional[WorkflowPhaseStatus] = None
    artifacts: list[SessionArtifactInfo] = []


class UpdateArtifactRequest(BaseModel):
    """아티팩트 수정 요청."""

    content: str


class AddAnnotationRequest(BaseModel):
    """인라인 주석 추가 요청."""

    line_start: int = Field(..., ge=1)
    line_end: Optional[int] = Field(None, ge=1)
    content: str = Field(..., min_length=1)
    annotation_type: AnnotationType = "comment"


class UpdateAnnotationRequest(BaseModel):
    """주석 상태 업데이트 요청."""

    status: Literal["resolved", "dismissed"]


class ApprovePhaseRequest(BaseModel):
    """Phase 승인 요청."""

    feedback: Optional[str] = None


class RequestRevisionRequest(BaseModel):
    """수정 요청."""

    feedback: str = Field(..., min_length=1)
